### ISSUE TITLE
Ability to tag an alert based on MITRE ATT&CK Framework

### ISSUE THREAD

**Issue by qjerome** [0xrawsec/gene#4]
opened by @michaelhidalgo in WHIDS project but moved here because more appropriate:

With MITRE ATT&CK becoming a standard way to classify an attacker action (aka technique) across the cyber kill change, it would be good to be able that, when a rule matches, be able to enrich that rule with the actual MITRE ATT&CK Tactic and Technique being used.

This has several benefits opening the door for doing correlation and eventually be able to draw (via some sort of dot notation graphs) the attacker objectives.

For example, this rule:

``` {
  "Name": "MaliciousLsassAccess",
  "Tags": ["Mimikatz", "Credentials", "Lsass"],
  "Meta": {
    "EventIDs": [10],
    "Channels": ["Microsoft-Windows-Sysmon/Operational"],
    "Computers": [],
    "Traces": [],
    "Criticality": 10,
    "Author": "0xrawsec"
  },
  "Matches": [
    "$ct: CallTrace ~= 'UNKNOWN'",
    "$lsass: TargetImage ~= '(?i:\\\\lsass\\.exe$)'"
  ],
  "Condition": "$lsass and $ct"
}```


Potentially matches [T1003](https://attack.mitre.org/techniques/T1003/)  Credential Dumping, so it would be nice to have this 

`

{
  "Event": {
    "EventData": {
      "CallTrace": "C:\\Windows\\SYSTEM32\\ntdll.dll+4bf9a|C:\\Windows\\system32\\KERNELBASE.dll+189b7|UNKNOWN(00000000259123BC)",
      "GrantedAccess": "0x1410",
      "SourceImage": "C:\\Windows\\system32\\WindowsPowerShell\\v1.0\\powershell.exe",
      "SourceProcessGUID": "{49F1AF32-DD18-5A72-0000-0010042C0A00}",
      "SourceProcessId": "2248",
      "SourceThreadId": "3308",
      "TargetImage": "C:\\Windows\\system32\\lsass.exe",
      "TargetProcessGUID": "{49F1AF32-DB3B-5A72-0000-001013690000}",
      "TargetProcessId": "492",
      "UtcTime": "2018-02-01 11:24:53.331"
    },
    "GeneInfo": {
      "Criticality": 10,
      "Signature": [
        "MaliciousLsassAccess"
      ]
      "ATT&CK Tactic"   : "Credential Access",
      "ATT&CK Technique : "T1003",
      "ATT&CK Reference": "https://attack.mitre.org/techniques/T1003/"
    },
    "System": {
        "Classical Windows Event System Section": "..."
    }
  }
}`

**Comment by qjerome** [comment]
On my way on implementing this

**Comment by michaelhidalgo** [comment]
Thanks @qjerome , so when the rule is defined, could we also setup the MITRE ATT$CK attributes from the rule definition?

**Comment by qjerome** [comment]
Of course, the MITRE ATT&CK attributes will be defined in the rules.
I am still testing the thing but the way I chose to encode this information in the rule is like this:
```json
{
  "Name": "ShouldMatch",
  "Meta": {
    "Channels": [
      "Microsoft-Windows-Sysmon/Operational"
    ],
    "ATT&CK": {
      "Tactic": "Blow everything up",
      "Technique": "T666",
      "Reference": "https://attack.mitre.org/"
    }
  },
  "Matches": [
    "$a: Hashes ~= SHA1=65894B0162897F2A6BB8D2EB13684BF2B451FDEE,"
  ],
  "Condition": "$a"
}
```
Does it make sense to you?

**Comment by michaelhidalgo** [comment]
Yes, absolutely makes sense, thanks!

**Comment by qjerome** [comment]
Hi @michaelhidalgo 

I slightly changed the format of the rules because I realized that a single rule can actually cover more than one technique. The example of the Suspicious Lsass access is a good illustration of this since it matches several techniques. So the rule format will likely be as such:
```json
{
	"Name": "ShouldMatch",
	"Meta": {
		"Channels": ["Microsoft-Windows-Sysmon/Operational"],
		"ATT&CK": [
			{
				"ID": "T666",
				"Tactic": "Blow everything up",
				"Reference": "https://attack.mitre.org/"
			},
			{
				"ID": "S4242",
				"Description": "Super nasty software",
				"Reference": "https://attack.mitre.org/"
			}
		]	
		},
	"Matches": [
		"$a: Hashes ~= 'SHA1=65894B0162897F2A6BB8D2EB13684BF2B451FDEE,'"
		],
	"Condition": "$a"
	}
```

**Comment by michaelhidalgo** [comment]
That is quite interesting @qjerome and that is one case I did not think of. I was actually thinking on an scenario like this:

![mitre](https://user-images.githubusercontent.com/1732035/57935241-9c991000-787e-11e9-93d0-cd9cb5e04de7.jpg)

This is the same technique but happening at different stages of the cyber kill chain (aka tactics). I would assume that involves separate rules to actually have more context on what to look for in order to properly tag a technique.

Also not sure if we should be that granular in terms of software or even groups, meaning that being able to tag the technique itself will give a lot of background to an analyst.

**Comment by qjerome** [comment]
What I meant is that there is not always a 1 to 1 relationship between a Gene rule and a technique described in the ATT&CK matrix. I rather think that a Gene rule at least may correspond to a technique but can match several techniques. I explain myself with a concrete example :). Let's assume the following, I decide to create a rule that catches any Sysmon process access (event ID 10) with write privilege to the foreign process. In this case the rule would correspond to several techniques (i.e. Process Hollowing (T1083), Processdoppleganging (T1186), Process Injection (T1055) and maybe others).

**Comment by michaelhidalgo** [comment]
Hey @qjerome , I guess it depends how granular one can be at the rule level using some context it could be possible to determine if the technique is a Process Hollowing or Process Injection. Unfortunately, MITRE ATT&CK is not that prescriptive and it requires to read all the research from the references and understand that "X group was observed using  Y tool to do Process Hollowing".

Context is necesary to avoid a high rate of false positives: for example, I have seen tools that, if the command line argument contains Powershell, they flag it with a technique so it is more a tagging than actual detection.

I'm also not sure if there is always a 1:1 relationship between a gene rule and a ATT&CK technique, probably it is if we can do fine tuning and context alike detections.

Note that there is another corner case where the Technique falls into one or more tactics (there are some techniques that appear 3 times in different stages). It's a different use case, but similar in the sense that requires probably more context to be able to understand where in the kill chain is happening.

**Comment by qjerome** [comment]
Implemented in v1.5.0
